modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/rel/CorrelatedNestedLoopJoinExecutionTest.java

```diff
@@ -0,0 +1,315 @@
+                ctx,
+                (r1, r2) -> Objects.equals(hnd.get(2, r1), hnd.get(0, r2)),
+                Set.of(new CorrelationId(0)),
+                joinType,
+                hnd.factory(rightRowSchema),
+                hnd.factory(outRowSchema)
+        );
+
+        join.register(Arrays.asList(left, right));
+
+        FilterNode<Object[]> filter = new FilterNode<>(ctx, r -> true);
+        filter.register(join);
+
+        RootNode<Object[]> root = new RootNode<>(ctx);
+        root.register(filter);
+
+        Object[][] result = StreamSupport.stream(Spliterators.spliteratorUnknownSize(root, Spliterator.ORDERED), false)
+                .toArray(Object[][]::new);
+
+        if (joinType == INNER) {
+            Object[][] expected = {
+                    {0, "Igor", 1, 1, "Core"},
+                    {1, "Roman", 2, 2, "SQL"},
+                    {3, "Alexey", 1, 1, "Core"}
+            };
+
+            assert2DimArrayEquals(expected, result);
+        } else if (joinType == LEFT) {
+            Object[][] expected = {
+                    new Object[]{0, "Igor", 1, 1, "Core"},
+                    new Object[]{1, "Roman", 2, 2, "SQL"},
+                    new Object[]{2, "Ivan", null, null, null},
+                    new Object[]{3, "Alexey", 1, 1, "Core"}
+            };
+
+            assert2DimArrayEquals(expected, result);
+        } else {
+            throw new IllegalStateException("Unexpected join type: " + joinType);
+        }
+    }
+
+    private static void assert2DimArrayEquals(Object[][] expected, Object[][] actual) {
+        assertEquals(expected.length, actual.length, "expected length: " + expected.length + ", actual length: " + actual.length);
+
+        Arrays.sort(actual, Comparator.comparing(r -> (int) r[0]));
+
+        int length = expected.length;
+        for (int i = 0; i < length; ++i) {
+            Object[] exp = expected[i];
+            Object[] act = actual[i];
+
+            assertEquals(exp.length, act.length, "expected length: " + exp.length + ", actual length: " + act.length);
+            assertArrayEquals(exp, act);
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = JoinRelType.class, mode = Mode.INCLUDE, names = {"INNER", "LEFT"})
+    void joinWithDifferentBufferSize(JoinRelType joinType) {
+        int buffSize = 1;
+        validateJoin(executionContext(buffSize), joinType, 0, 0);
+        validateJoin(executionContext(buffSize), joinType, 0, 1);
+        validateJoin(executionContext(buffSize), joinType, 0, 10);
+        validateJoin(executionContext(buffSize), joinType, 1, 0);
+        validateJoin(executionContext(buffSize), joinType, 1, 1);
+        validateJoin(executionContext(buffSize), joinType, 1, 10);
+        validateJoin(executionContext(buffSize), joinType, 10, 0);
+        validateJoin(executionContext(buffSize), joinType, 10, 1);
+        validateJoin(executionContext(buffSize), joinType, 10, 10);
+
+        buffSize = Commons.IN_BUFFER_SIZE;
```

Comment:
we have significant small rows than IN_BUFFER_SIZE, probably we need to test with rows equal to IN_BUFFER_SIZE ? wdyt ?

Reply:
i miss it, thanks !